### PR TITLE
[C++] Use malloc/free instead of k_malloc/k_free in operator new/delete

### LINKED ISSUE #18990: C++ New allocates memory from kernel heap
**Describe the bug**

The current Zephyr C++ `new` implementation allocates memory from the kernel heap instead of using the same heap used by `malloc`. 

In case a system would be configured with none, or a small, kernel heap and a large libc heap the call to `new` will fail even when sufficient memory is available in the libc heap.

**To Reproduce**
- Problem was hit on private project.
- Todo: reproduce with libcxx tests

**Expected behavior**

`new`/`delete` should use same memory pool as `malloc`/`free`

**Impact**

None useable system due to memory allocation failure.

**Environment**
 - OS: Linux
 - Toolchain: Zephyr
 - f9c94d49e84def297c29f800b8733e5763b8d002 (2.0-rc3)

**Additional context**

PR (#18464) was created to provide the option to disable Zephyr `new`/`free` implementation and falback to libcxx implementation which uses `malloc` and `free` internally.

One open question from the code review on PR #18464 is why zephyr provides its own implementation of `new`/`free` that specifically depends on kernel heap

### PR DESCRIPTION
**Pull request by vanwinkeljan**
Use malloc/free instead of k_malloc/k_free in operator new/delete and extended test cases to cover new/delete.

Fixes: #18990

### CODE REVIEW

File: tests/application_development/cpp/testcase.yaml [hunk 1]
```diff
@@ -1,5 +1,5 @@
 tests:
   misc.app_dev.cpp:
     arch_exclude: posix
-    build_only: true
```
[the remark this answers is not in the document]
Comment (vanwinkeljan) [reply tests/application_development/cpp/testcase.yaml]: I added a test case to cover new/delete and this onde can run in CI, before this change the test case did not contained any runnable test and was only build.

**Comment by aescolar** [comment]
@vanwinkeljan What is the motivation for doing this change?

**Comment by vanwinkeljan** [comment]
@aescolar please see issue  #18990 for more background

**Comment by pabigot** [comment]
This isn't what I expected.  Please check my understanding:

* Currently `ZEPHYR_CPLUSPLUS` is defaulted off, and applies only when `MINIMAL_LIBC` is disabled and `LIB_CPLUSPLUS` is enabled.  AFAICT `ZEPHYR_CPLUSPLUS` would be desirable only when using a libc that doesn't support C++ even though it provides a libc++.  (??  I'm not sure how that would even work, but I think it was a path to retaining old behavior, so fine.)  Its effect is to support new/delete and virtual functions.
* Proposed `ZEPHYR_NEW_DELETE` is defaulted on when `LIBCPLUSPLUS` is disabled, requires `MINIMAL_LIBC` is disabled or supports malloc, and is allowed when `LIB_CPLUSPLUS` is enabled.  It only overrides new/delete, and overriding virtual support is always done with `LIB_CPLUSPLUS` is disabled.

At this point I have no clue of how all the different C++ and libc Kconfig features interact.  I'm particularly concerned here that it might become easy for somebody to replace new/delete in a newlib-based libc++ (which already implements them using malloc), which would be gratuitous and risky.

That mess pre-dates this PR, but I'm uncomfortable with changing it at this point.  For the purposes of settling the basic need in #18990 to get rid of using kernel memory for C++ objects is it not sufficient to just replace the `k_malloc(free)` preprocessor-protected calls in `cpp_new.cpp` with unprotected `malloc(free)` calls?  And leave the configuration infrastructure unchanged?

**Comment by vanwinkeljan** [comment]
> I'm particularly concerned here that it might become easy for somebody to replace new/delete in a newlib-based libc++ (which already implements them using malloc), which would be gratuitous and risky.

We could just drop the option of using the zephyr version of new/delete if a libc++ is used (it uses malloc internally). Actually this is maybe the most correct and clean way forward and actually the Kconfig could be dropped in this way.

Shall I update the PR to get a feeling how it would look like?

**Comment by vanwinkeljan** [comment]
@pabigot @andrewboie Another surprise is the failing CI, it seems that virtual destructors are calling `operator delete`. With the current proposal this requires a libc heap.

File: tests/application_development/cpp/testcase.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (andrewboie) [reply tests/application_development/cpp/testcase.yaml]: right but why did you exclude qemu_x86_coverage

**Comment by andrewboie** [comment]
> We could just drop the option of using the zephyr version of new/delete if a libc++ is used (it uses malloc internally).

Yes, why would we not want to do this?

Maybe this is historical cruft but we should not be providing our own versions of these functions if the C library already provides them!

File: tests/application_development/cpp/testcase.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (vanwinkeljan) [reply tests/application_development/cpp/testcase.yaml]: For some reason qemu_x86_coverage hangs when running with c++.

[the remark this answers is not in the document]
Comment (andrewboie) [reply tests/application_development/cpp/testcase.yaml]: need a bug filed on this, since this is an indicator that gathering code coverage will fail

**Review by pabigot — APPROVED** [review]

File: tests/application_development/cpp/testcase.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (vanwinkeljan) [reply tests/application_development/cpp/testcase.yaml]: created bug report #20729, looks like our own coverage code ends up in a endless loop

**Review by tyrantlizard77 — APPROVED** [review]

**Comment by dleach02** [comment]
@vanwinkeljan, yes we should merge this to clear the medium bug.

**Pull request merged**